### PR TITLE
Request via unix domain socket

### PR DESCRIPTION
**Pull request by vdemedes**
This PR adds an ability to send http(s) requests via unix domain socket.

### Known use-cases:

- Docker
- fleet

### Usage

To send a request via unix domain socket, use the following URL scheme (borrowed from [request](https://www.npmjs.com/package/request#unix-domain-sockets) module): `protocol://unix:/path/to/socket:/request/path`, where:

- `protocol` - one of: `http`, `https`
- `/path/to/socket` - an absolute path to a unix socket
- `/request/path` - http request path

```js
var got = require('got');

got('http://unix:/var/run/docker.sock:/');
```

All the existing functionality is preserved.

### Tests

Tests for this PR are in `test/test-unix-socket.js`.

Let me know what you think ;)

### CODE REVIEW

**Comment by sindresorhus** [comment]
:+1: 

@vdemedes Can you add some docs? Similar to https://github.com/request/request#unix-domain-sockets Preferably with real use-cases.

@floatdrop What do you think?

**Comment by vdemedes** [comment]
@sindresorhus Yeah, sure, will do.

File: test/test-unix-socket.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+test('setup', function (t) {
+	s.listen(socketPath, function () {
+		t.end();
+	});
+});
+
+test('request via unix socket', function (t) {
+	// borrow unix domain socket url format from request module
+	var url = format('http://unix:%s:%s', socketPath, '/');
```
Comment (sindresorhus) [inline test/test-unix-socket.js]: Can you add a test to make sure `unix:%s:%s', socketPath, '/'` works too? We support protocol-less URLs for normal URLs.

Comment (vdemedes) [inline test/test-unix-socket.js]: Yeah, ok ;)

**Comment by vdemedes** [comment]
@sindresorhus done ;)

File: readme.md [hunk 2]
```diff
@@ -201,6 +201,25 @@ got('todomvc.com', {
...
+
+- `PROTOCOL` - `http` or `https`
+- `SOCKET` - absolute path to a unix domain socket, e.g. `/var/run/docker.sock`
+- `PATH` - request path, e.g. `/v2/keys`
+
+Example:
+
+```js
+got('http://unix:/var/run/docker.sock:/containers/json');
+```
```
Comment (sindresorhus) [inline readme.md]:
```js
got('http://unix:/var/run/docker.sock:/containers/json');

// or
got('unix:/var/run/docker.sock:/containers/json');
```

File: readme.md [hunk 3]
```diff
@@ -201,6 +201,25 @@ got('todomvc.com', {
 }, function () {});
 ```
 
+### Unix Domain Sockets
+
+Requests can also be sent via unix domain sockets. Use the following URL scheme: `PROTOCOL://unix:SOCKET:PATH`.
+
+- `PROTOCOL` - `http` or `https`
```
Comment (sindresorhus) [inline readme.md]:
```
- `PROTOCOL` - `http` or `https` *(optional)*
```

File: readme.md [hunk 4]
```diff
@@ -201,6 +201,25 @@ got('todomvc.com', {
 }, function () {});
 ```
 
+### Unix Domain Sockets
+
+Requests can also be sent via unix domain sockets. Use the following URL scheme: `PROTOCOL://unix:SOCKET:PATH`.
```
Comment (sindresorhus) [inline readme.md]: Can you linkify `unix domain sockets` to an article/page where the user can read more about unix domain sockets?

File: readme.md [hunk 5]
```diff
@@ -201,6 +201,25 @@ got('todomvc.com', {
...
+
+Example:
+
+```js
+got('http://unix:/var/run/docker.sock:/containers/json');
+```
+
+Use-cases:
+
+- Docker API (/var/run/docker.sock)
```
Comment (sindresorhus) [inline readme.md]: Linkify to some Docker docs about it?

File: readme.md [hunk 6]
```diff
@@ -201,6 +201,25 @@ got('todomvc.com', {
...
+- fleet API  (/var/run/fleet.sock)
```
Comment (sindresorhus) [inline readme.md]: Same as above.

**Comment by vdemedes** [comment]
@sindresorhus Read all notes, fix is coming very soon, thx ;)

File: index.js [hunk 7]
```diff
@@ -28,6 +28,23 @@ function requestAsEventEmitter(opts) {
 	var get = function (opts) {
 		var fn = opts.protocol === 'https:' ? https : http;
 
+		// check for unix domain socket
+		if (opts.hostname === 'unix') {
```
Comment (floatdrop) [inline index.js]: Is this check belongs here? Why not in `normalizeAgruments` (redirect can return unix socket in headers?)?

**Comment by floatdrop** [comment]
:+1: Thank you!

**Pull request closed without merging**